api/stripe.go

```diff
@@ -0,0 +1,313 @@
+		s, err := sub.Get(hasSub.Sub, nil)
+		if err != nil {
+			api.staticLogger.Debugln("Failed to fetch sub:", err)
+			return
+		}
+		err = api.processStripeSub(req.Context(), s)
+		if err != nil {
+			api.staticLogger.Debugln("Failed to process sub:", err)
+		}
+	}
+
+	api.WriteSuccess(w)
+}
+
+// stripePricesHandler returns a list of plans and prices.
+func (api *API) stripePricesHandler(w http.ResponseWriter, req *http.Request, _ httprouter.Params) {
+	api.staticLogger.Tracef("Processing request: %+v", req)
+	var sPrices []stripePrice
+	params := &stripe.PriceListParams{Active: &True}
+	params.AddExpand("data.product")
+	params.Filters.AddFilter("limit", "", "1000")
+	i := price.List(params)
+	for i.Next() {
+		p := i.Price()
+		sp := stripePrice{
+			ID:          p.ID,
+			Name:        p.Product.Name,
+			Description: p.Product.Description,
+			Tier:        stripePrices()[p.ID],
+			Price:       float64(p.UnitAmount) / 100,
+			Currency:    string(p.Currency),
+			StripeID:    p.ID,
+			ProductID:   p.Product.ID,
+			LiveMode:    p.Livemode,
+		}
+		sPrices = append(sPrices, sp)
+	}
+	api.WriteJSON(w, sPrices)
+}
+
+// readStripeEvent reads the event from the request body and verifies its
+// signature.
+func readStripeEvent(w http.ResponseWriter, req *http.Request) (*stripe.Event, int, error) {
+	const MaxBodyBytes = int64(65536)
+	req.Body = http.MaxBytesReader(w, req.Body, MaxBodyBytes)
+	payload, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		err = errors.AddContext(err, "error reading request body")
+		return nil, http.StatusServiceUnavailable, err
+	}
+	// Read the event and verify its signature.
+	event, err := webhook.ConstructEvent(payload, req.Header.Get("Stripe-Signature"), os.Getenv("STRIPE_WEBHOOK_SECRET"))
+	if err != nil {
+		return nil, http.StatusBadRequest, err
+	}
+	return &event, http.StatusOK, nil
+}
+
+// processStripeSub reads the information about the user's subscription and
+// adjusts the user's record accordingly.
+func (api *API) processStripeSub(ctx context.Context, s *stripe.Subscription) error {
+	api.staticLogger.Traceln("Processing subscription:", s.ID)
+	// Get all active subscriptions for this customer. There should be only one
+	// (or none) but we'd better check.
+	it := sub.List(&stripe.SubscriptionListParams{
+		Customer: s.Customer.ID,
+		Status:   string(stripe.SubscriptionStatusActive),
+	})
+	u, err := api.staticDB.UserByStripeID(ctx, s.Customer.ID)
+	if err != nil {
+		return errors.AddContext(err, "failed to fetch user from DB based on subscription info")
+	}
+	// Set the default values.
+	u.Tier = database.TierFree
+	u.SubscribedUntil = time.Time{}
+	u.SubscriptionStatus = ""
+	u.SubscriptionCancelAt = time.Time{}
+	u.SubscriptionCancelAtPeriodEnd = false
+	var latestCreated time.Time
+	var latestSub string
+	// Pick the latest active plan and set the user's tier based on that.
+	subs := it.SubscriptionList().Data
+	for _, subsc := range subs {
```

Comment:
Why do you loop over all subs and decorate the user potentially multiple times? 
Seems more logical to find the most recent sub and then decorate it once?

I'm also slightly confused on why we reset the user's subscription fields to default,
before assigning it the new ones.

Reply:
The main goal was to avoid the `else` in `if <has sub> <decorate with sub> else <decorate with free>`. But I don't mind changing it. Will F/U in #38.